src/analyses/base.ml

```diff
@@ -1997,6 +1997,20 @@ struct
     let st' = invalidate ~deep:false ~ctx (Analyses.ask_of_ctx ctx) gs st shallow_addrs in
     invalidate ~deep:true ~ctx (Analyses.ask_of_ctx ctx) gs st' deep_addrs
 
+  let check_free_of_non_heap_mem ctx special_fn ptr =
+    let points_to_set = ctx.ask (Queries.MayPointTo ptr) in
+    begin try
+        let exists_non_heap_var =
+          (* elements throws Unsupported if the points-to set is top *)
+          Queries.LS.elements points_to_set
+          |> List.map fst
+          |> List.exists (fun var -> not (ctx.ask (Queries.IsHeapVar var)))
+        in
```

Comment:
This is in base, so a query is kind of roundabout. Other parts of base just use `eval_rv_address` or something to look up points-to sets and avoid the unnecessary back-and-forth conversions with `Queries.LS`.